File: src/GitOpen.Domain/Diff/DiffSpec.cs
using GitOpen.Domain.Commits;

namespace GitOpen.Domain.Diff;

public abstract record DiffSpec
{
    public sealed record CommitVsParent(CommitSha CommitSha) : DiffSpec;
    public sealed record CommitVsCommit(CommitSha From, CommitSha To) : DiffSpec;
    public sealed record IndexVsHead : DiffSpec;
    public sealed record WorkingTreeVsIndex : DiffSpec;
}

File: lib/domain/diff/diff_spec.dart
import 'package:equatable/equatable.dart';

import 'package:gitopen/domain/commits/commit_sha.dart';

sealed class DiffSpec extends Equatable {
  const DiffSpec();
}

final class DiffSpecCommitVsParent extends DiffSpec {

  const DiffSpecCommitVsParent(this.commitSha);
  final CommitSha commitSha;

  @override
  List<Object?> get props => [commitSha];
}

final class DiffSpecCommitVsCommit extends DiffSpec {

  const DiffSpecCommitVsCommit(this.from, this.to);
  final CommitSha from;
  final CommitSha to;

  @override
  List<Object?> get props => [from, to];
}

final class DiffSpecIndexVsHead extends DiffSpec {
  const DiffSpecIndexVsHead();

  @override
  List<Object?> get props => const [];
}

final class DiffSpecWorkingTreeVsIndex extends DiffSpec {
  const DiffSpecWorkingTreeVsIndex();

  @override
  List<Object?> get props => const [];
}
